File: TwoPointer/FourSum.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class FourSum {
    public static void main(String[] args) {
        FourSum solution = new FourSum();
        int[] nums = { 1, 0, -1, 0, -2, 2 };
        int target = 0;
        List<List<Integer>> ans = solution.input(nums, target);
        System.out.println(ans);
    }

    public List<List<Integer>> input(int[] nums, int target) {

        List<List<Integer>> quadruplets = new ArrayList<>();
        int n = nums.length;
        Arrays.sort(nums);
        for (int i = 0; i < n - 3; i++) {
            if (i > 0 && nums[i] == nums[i - 1])
                continue;
            for (int j = i + 1; j < n - 2; j++) {
                if (j > i + 1 && nums[j] == nums[j - 1])
                    continue;
                int left = j + 1;
                int right = n - 1;
                while (left < right) {
                    long sum = (long) nums[i] + nums[j] + nums[left] + nums[right];
                    if (sum < target) {
                        left++;
                    } else if (sum > target) {
                        right--;
                    } else {
                        quadruplets.add(Arrays.asList(nums[i], nums[j], nums[left], nums[right]));
                        while (left < right && nums[left] == nums[left + 1]) {
                            left++;
                        }
                        while (left < right && nums[right] == nums[right - 1]) {
                            right--;
                        }
                        left++;
                        right--;
                    }
                }
            }
        }
        return quadruplets;
    }
}

File: TwoPointer/FourSum.js
var fourSum = function(nums, target) {
    let qlist = [];
    nums.sort((a, b) => a - b); // 对数组进行排序

    for (let i = 0; i < nums.length - 3; i++) {
        if (i > 0 && nums[i] === nums[i - 1]) {
            continue; // 跳过重复元素
        }

        for (let j = i + 1; j < nums.length - 2; j++) {
            if (j > i + 1 && nums[j] === nums[j - 1]) {
                continue; // 跳过重复元素
            }

            let left = j + 1;
            let right = nums.length - 1;

            while (left < right) {
                const sum = nums[i] + nums[j] + nums[left] + nums[right];

                if (sum === target) {
                    qlist.push([nums[i], nums[j], nums[left], nums[right]]);
                    while (left < right && nums[left] === nums[left + 1]) {
                        left++;
                    }
                    while (left < right && nums[right] === nums[right - 1]) {
                        right--;
                    }
                    left++;
                    right--;
                } else if (sum < target) {
                    left++;
                } else {
                    right--;
                }
            }
        }
    }

    return qlist;
};
// var FourSum=function(nums,target){
//     let qlist=[];
//     for(let i=0;i<nums.length-3;i++){
//         if(i>0&&nums[i]===nums[i-1]){
//             continue;
//         }
//         for(let j=i;j<nums.length-2;j++){
//             if(j>i&&nums[j]===nums[j-1]){
//                 continue;
//             }
//             let left=j+1;
//             let right=nums.length-1;
//             while(left<right){
//                 let sum=nums[i]+nums[j]+nums[left]+nums[right];
//                 if(sum<target){
//                     left++;
//                 }else if(sum>target){
//                     right--;
//                 }else{
//                     qlist.push([nums[i],nums[j],nums[left],nums[right]]);
//                     while(left<right&&nums[left]===nums[left+1]){
//                         left++;
//                     };
//                     while(left<right&&nums[right]===nums[right-1]){
//                         right--;
//                     }
//                     left++;
//                     right--;
//                 }
//             }
//         }
      
//     }  return qlist;
// };
const nums=[1,0,-1,0,-2,2];
const target=0;
const result=FourSum(nums,target);
console.log(result);